From 0f1d2471dac3128cb3c580e9064484383c90b9fd Mon Sep 17 00:00:00 2001
From: Mohsen Dolaty <mohsen.dolaty@amd.com>
Date: Thu, 23 Feb 2023 14:21:14 -0600
Subject: [PATCH 1/1] linux-aspeed: Add Turin Lenovo UBM

Add UBM backplane devices for Turin Volcano and
Purico platforms (MUXs, PSOCs, FRUs)
Clean up ENABLE_I2C_APML definition

Signed-off-by: Mohsen Dolaty <mohsen.dolaty@amd.com>
---
 arch/arm/boot/dts/aspeed-bmc-amd-purico.dts  | 181 +++++++++++++++++-
 arch/arm/boot/dts/aspeed-bmc-amd-volcano.dts | 187 +++++++++++++++++--
 2 files changed, 352 insertions(+), 16 deletions(-)

diff --git a/arch/arm/boot/dts/aspeed-bmc-amd-purico.dts b/arch/arm/boot/dts/aspeed-bmc-amd-purico.dts
index fa66061a3aea..be715e0e2bc1 100644
--- a/arch/arm/boot/dts/aspeed-bmc-amd-purico.dts
+++ b/arch/arm/boot/dts/aspeed-bmc-amd-purico.dts
@@ -16,6 +16,22 @@ aliases {
 		serial4 = &uart5;
 		ethernet0 = &mac3;
 		ethernet1 = &mac2;
+		i2c240 = &channel_10_10; //U63
+		i2c241 = &channel_10_11;
+		i2c242 = &channel_10_12;
+		i2c243 = &channel_10_13;
+		i2c244 = &channel_10_14;
+		i2c245 = &channel_10_15;
+		i2c246 = &channel_10_16;
+		i2c250 = &BP1;			//U134
+		i2c251 = &BP2;
+		i2c252 = &BP3;
+		i2c255 = &BP1_Crtl0;	//U6
+		i2c256 = &BP1_Crtl1;
+		i2c257 = &BP2_Crtl0;
+		i2c258 = &BP2_Crtl1;
+		i2c259 = &BP3_Crtl0;
+		i2c260 = &BP3_Crtl1;
 	};

 	chosen {
@@ -364,9 +380,9 @@ i2cswitch@70 {
 	};

 	// SCM brd_id, Purico brd_id, CLK
-	i2cswitch@71 {
+	i2cswitch@73 {
 		compatible = "nxp,pca9548";
-		reg = <0x71>;
+		reg = <0x73>;
 		#address-cells = <1>;
 		#size-cells = <0>;

@@ -387,6 +403,165 @@ scmbrdeeprom@51 {
 		};
 	};
 #endif
+	//Lenovo Backplane
+	i2cswitch@75 {
+		compatible = "nxp,pca9548";
+		reg = <0x75>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		channel_10_10: i2c@0 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0>;
+		};
+
+		channel_10_11: i2c@1 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <1>;
+		};
+
+		channel_10_12: i2c@2 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <2>;
+		};
+
+		channel_10_13: i2c@3 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <3>;
+
+			// BP
+			i2cswitch@70 {
+				compatible = "nxp,pca9545";
+				#address-cells = <1>;
+				#size-cells = <0>;
+				reg = <0x70>;
+
+				//  channel 0 connected to BP1
+				BP1: i2c@0 {
+					#address-cells = <1>;
+					#size-cells = <0>;
+					reg = <0>;
+
+					bp1_eeprom: eeprom@50 {
+						compatible = "atmel,24c32";
+						reg = <0x50>;
+						pagesize = <32>;
+					};
+
+					i2cswitch@71 {
+						compatible = "nxp,pca9545";
+						#address-cells = <1>;
+						#size-cells = <0>;
+						reg = <0x71>;
+
+						//  channel 0 connected to PSOC0
+						BP1_Crtl0: i2c@0 {
+							#address-cells = <1>;
+							#size-cells = <0>;
+							reg = <0>;
+						};
+
+						//  channel 1 connected to PSOC1
+						BP1_Crtl1: i2c@1 {
+							#address-cells = <1>;
+							#size-cells = <0>;
+							reg = <1>;
+						};
+					};
+				};
+
+				//  channel 1 connected to BP2
+				BP2: i2c@1 {
+					#address-cells = <1>;
+					#size-cells = <0>;
+					reg = <1>;
+
+					bp2_eeprom: eeprom@50 {
+						compatible = "atmel,24c32";
+						reg = <0x50>;
+						pagesize = <32>;
+					};
+
+					i2cswitch@71 {
+						compatible = "nxp,pca9545";
+						#address-cells = <1>;
+						#size-cells = <0>;
+						reg = <0x71>;
+
+						//  channel 0 connected to PSOC0
+						BP2_Crtl0: i2c@0 {
+							#address-cells = <1>;
+							#size-cells = <0>;
+							reg = <0>;
+						};
+
+						//  channel 1 connected to PSOC1
+						BP2_Crtl1: i2c@1 {
+							#address-cells = <1>;
+							#size-cells = <0>;
+							reg = <1>;
+						};
+					};
+				};
+
+				//  channel 2 connected to BP3
+				BP3: i2c@2 {
+					#address-cells = <1>;
+					#size-cells = <0>;
+					reg = <2>;
+
+					bp3_eeprom: eeprom@50 {
+						compatible = "atmel,24c32";
+						reg = <0x50>;
+						pagesize = <32>;
+					};
+
+					i2cswitch@71 {
+						compatible = "nxp,pca9545";
+						#address-cells = <1>;
+						#size-cells = <0>;
+						reg = <0x71>;
+
+						//  channel 0 connected to PSOC0
+						BP3_Crtl0: i2c@0 {
+							#address-cells = <1>;
+							#size-cells = <0>;
+							reg = <0>;
+						};
+
+						//  channel 1 connected to PSOC1
+						BP3_Crtl1: i2c@1 {
+							#address-cells = <1>;
+							#size-cells = <0>;
+							reg = <1>;
+						};
+					};
+				};
+			};
+		};
+
+		channel_10_14: i2c@4 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <4>;
+		};
+
+		channel_10_15: i2c@5 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <5>;
+		};
+
+		channel_10_16: i2c@6 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <6>;
+		};
+	};
 };

 // BCM5720 LOM
@@ -764,7 +939,7 @@ rcd_3_5: rcd@5d,2C000000005 {
         };
 };

-#define ENABLE_I2C_APML (1)
+#define ENABLE_I2C_APML
 #ifdef ENABLE_I2C_APML

 &i2c2 {
diff --git a/arch/arm/boot/dts/aspeed-bmc-amd-volcano.dts b/arch/arm/boot/dts/aspeed-bmc-amd-volcano.dts
index fe7a712bb2d8..0dd5fce7f560 100644
--- a/arch/arm/boot/dts/aspeed-bmc-amd-volcano.dts
+++ b/arch/arm/boot/dts/aspeed-bmc-amd-volcano.dts
@@ -16,6 +16,22 @@ aliases {
 		serial4 = &uart5;
 		ethernet0 = &mac3;
 		ethernet1 = &mac2;
+		i2c240 = &channel_10_10; //U63
+		i2c241 = &channel_10_11;
+		i2c242 = &channel_10_12;
+		i2c243 = &channel_10_13;
+		i2c244 = &channel_10_14;
+		i2c245 = &channel_10_15;
+		i2c246 = &channel_10_16;
+		i2c250 = &BP1;			//U134
+		i2c251 = &BP2;
+		i2c252 = &BP3;
+		i2c255 = &BP1_Crtl0;	//U6
+		i2c256 = &BP1_Crtl1;
+		i2c257 = &BP2_Crtl0;
+		i2c258 = &BP2_Crtl1;
+		i2c259 = &BP3_Crtl0;
+		i2c260 = &BP3_Crtl1;
 	};

 	chosen {
@@ -525,7 +541,7 @@ rcd_3_5: rcd@5d,2C000000005 {
 	};
 };

-#define ENABLE_I2C_APML (1)
+#define ENABLE_I2C_APML
 #ifdef ENABLE_I2C_APML

 &i2c2 {
@@ -810,9 +826,9 @@ &i2c10 {
 	status = "okay";
 #ifdef EEPROM_PROG_ENABLE
 	// SCM brd_id, Volcano brd_id, CLK
-	i2cswitch@71 {
+	i2cswitch@73 {
 		compatible = "nxp,pca9848";
-		reg = <0x71>;
+		reg = <0x73>;
 		#address-cells = <1>;
 		#size-cells = <0>;

@@ -832,23 +848,168 @@ scmbrdeeprom@51 {
 				reg = <0x51>;
 			};

-			// Usually the eeproms are at 0x50 and 0x51 address,
-			// but some systems (Rev A) have eeprom address 0x54
-			// SCM BRD ID eeprom
-			scmbrdeeprom@54 {
-				compatible = "microchip,24lc256","atmel,24c256";
-				reg = <0x54>;
-			};
 		};
 	};
-	// Midpalne, Raiser cards, SATA, etc
+#endif
+	// Midplane, Raiser cards, SATA, etc
 	i2cswitch@75 {
-		compatible = "nxp,pca9848";
+		compatible = "nxp,pca9548";
 		reg = <0x75>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+
+		channel_10_10: i2c@0 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0>;
+		};
+
+		channel_10_11: i2c@1 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <1>;
+		};
+
+		channel_10_12: i2c@2 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <2>;
+		};
+
+		channel_10_13: i2c@3 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <3>;
+
+			// BP
+			i2cswitch@70 {
+				compatible = "nxp,pca9545";
+				#address-cells = <1>;
+				#size-cells = <0>;
+				reg = <0x70>;
+
+				//  channel 0 connected to BP1
+				BP1: i2c@0 {
+					#address-cells = <1>;
+					#size-cells = <0>;
+					reg = <0>;
+
+					bp1_eeprom: eeprom@50 {
+						compatible = "atmel,24c32";
+						reg = <0x50>;
+						pagesize = <32>;
+					};
+
+					i2cswitch@71 {
+						compatible = "nxp,pca9545";
+						#address-cells = <1>;
+						#size-cells = <0>;
+						reg = <0x71>;
+
+						//  channel 0 connected to PSOC0
+						BP1_Crtl0: i2c@0 {
+							#address-cells = <1>;
+							#size-cells = <0>;
+							reg = <0>;
+						};
+
+						//  channel 1 connected to PSOC1
+						BP1_Crtl1: i2c@1 {
+							#address-cells = <1>;
+							#size-cells = <0>;
+							reg = <1>;
+						};
+					};
+				};
+
+				//  channel 1 connected to BP2
+				BP2: i2c@1 {
+					#address-cells = <1>;
+					#size-cells = <0>;
+					reg = <1>;
+
+					bp2_eeprom: eeprom@50 {
+						compatible = "atmel,24c32";
+						reg = <0x50>;
+						pagesize = <32>;
+					};
+
+					i2cswitch@71 {
+						compatible = "nxp,pca9545";
+						#address-cells = <1>;
+						#size-cells = <0>;
+						reg = <0x71>;
+
+						//  channel 0 connected to PSOC0
+						BP2_Crtl0: i2c@0 {
+							#address-cells = <1>;
+							#size-cells = <0>;
+							reg = <0>;
+						};
+
+						//  channel 1 connected to PSOC1
+						BP2_Crtl1: i2c@1 {
+							#address-cells = <1>;
+							#size-cells = <0>;
+							reg = <1>;
+						};
+					};
+				};
+
+				//  channel 2 connected to BP3
+				BP3: i2c@2 {
+					#address-cells = <1>;
+					#size-cells = <0>;
+					reg = <2>;
+
+					bp3_eeprom: eeprom@50 {
+						compatible = "atmel,24c32";
+						reg = <0x50>;
+						pagesize = <32>;
+					};
+
+					i2cswitch@71 {
+						compatible = "nxp,pca9545";
+						#address-cells = <1>;
+						#size-cells = <0>;
+						reg = <0x71>;
+
+						//  channel 0 connected to PSOC0
+						BP3_Crtl0: i2c@0 {
+							#address-cells = <1>;
+							#size-cells = <0>;
+							reg = <0>;
+						};
+
+						//  channel 1 connected to PSOC1
+						BP3_Crtl1: i2c@1 {
+							#address-cells = <1>;
+							#size-cells = <0>;
+							reg = <1>;
+						};
+					};
+				};
+			};
+		};
+
+		channel_10_14: i2c@4 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <4>;
+		};
+
+		channel_10_15: i2c@5 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <5>;
+		};
+
+		channel_10_16: i2c@6 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <6>;
+		};
 	};
-#endif
 };

 // 1G ROM
--
2.25.1